### ISSUE TITLE
ServiceStack.Text deserialization issue

### ISSUE THREAD

**Issue by BreakingPanda** [ServiceStack/Issues#636]
I found an issue working with this library.

1- I deserialize string formatted to C# object using ServiceStack.Text and it always work as expected:
```
        public static T Deserialize<T>(this string text)
        {
            using (JsConfig.With(emitLowercaseUnderscoreNames: true, 
                                 emitCamelCaseNames: false, 
                                 propertyConvention: PropertyConvention.Lenient))
            {
                return text.FromJson<T>();
            }
        }
```
As you can see, I create a scope when transforming string text to C#. 

2. My API successfully prepares the C# objects and I return it so ServiceStack can serialize it and send back a string back to consumer because my API is JSON REST API. 
I use ServiceStack library, and transformation is handled by ServiceStack and I have this setting in my AppHost:
```
        public override void Configure(Container container)
        {
            JsConfig.DateHandler = DateHandler.ISO8601;
            JsConfig.AlwaysUseUtc = true;
            JsConfig.EmitLowercaseUnderscoreNames = false;
            JsConfig.EmitCamelCaseNames = false;
            JsConfig.ExcludeDefaultValues = true;
             ....
}
```

My API returns this:
```
public class AddressSuggestion
    {
        public string StreetAddress { get; set; }
        public string FormattedAddress { get; set; }
        public string Postcode { get; set; }
        public string Suburb { get; set; }
        public string City { get; set; }
        public string State { get; set; }
        public string Country { get; set; }
        public string SearchResultId { get; set; }
        public int Index { get; set; }
        ....
    }
```
The issue is that, my JSON API works as expected for few hours and the output is as:

```
    {
        "StreetAddress": "6 NILE CL",
        "FormattedAddress": "6 Nile Cl, GERRINGONG  NSW  2534",
        "Postcode": "2534",
        "Suburb": "GERRINGONG",
        "State": "NSW",
        "SearchResultId": "ececedcb-804c-4c2f-9359-fa0270e24ce0--1",
        "Index": 1,
    }
```
And then when API enters peak time, my API returns this and never switched back to correct respond:
```
    {
        "street_address": "6 NILE CL",
        "formatted_address": "6 Nile Cl, GERRINGONG  NSW  2534",
        "postcode": "2534",
        "suburb": "GERRINGONG",
        "state": "NSW",
        "search_result_id": "ececedcb-804c-4c2f-9359-fa0270e24ce0--1",
        "index": 1,
    }
```
Having underscore in property name of response causes heaps issue for my other APIs in production. 

Any advise?

**Comment by mythz** [comment]
I'm most likely [going to need a repro](https://github.com/ServiceStack/Issues#servicestack-issues) in order to identify and resolve this issue. I'm assuming the static `JsConfig` properties are never modified after startup right?

**Comment by BreakingPanda** [comment]
1. Yes, I set JSCobfig globally in AppHost only.
2. This issue is hard to reproduce as I only found occurrences in production. Same API works nicely in low traffic environment. 

My suggestion is to review possibility of threading issue. I have seen source code for ServiceStack.Text and excessive use of static variable might be a reason for this thread unsafety issue.

**Comment by mythz** [comment]
I've already looked and I'm not seeing any issue which [calls this config accessor every time](https://github.com/ServiceStack/ServiceStack.Text/blob/master/src/ServiceStack.Text/JsConfig.cs#L581), if it's never modified after Startup the variable should be immutable and immune to race conditions. I'm assuming it's leaking out of the scope (i.e. using async within scope or scope not disposed properly), but without being able to reproduce it I wont be able to tell if it's due to an external environment condition that I wont be able to reproduce, hence why I need the repro, to be able to identify it first then verify that a fix resolves it.

**Comment by BreakingPanda** [comment]
1. If you look again how I create scope, I can assure you that scope is disposed properly + I don’t have any async call in a scope.
2. I have no code to change global setting except of AppHost startup (as I provided the code for it)
3. I fully understand how difficult it is to resolve/advise an issue without repro sample. I will see if I can give you a simple code BUT as mentioned before I CANNOT REPRO THIS EITHER in my local/test environment. This issue only occurs in production environment after few hours in peak time. To me , this is certainly a thread safety issue which can happen in a rare situation. So I am afraid we cannot really reproduce hence no solution!

I have a license and wonder what level of support I can expect if upgrade to premium.

Thanks for helping.

**Comment by mythz** [comment]
I've rewritten the internal configuration model to use "config snapshots" in [this commit](https://github.com/ServiceStack/ServiceStack.Text/commit/086a74be50c7bddc7abc1b556e54ab51d7654b8a) where the configuration uses either scoped configuration or if there's no scope it uses global configuration, i.e. it no longer fallsback to use Global configuration if a scope configuration wasn't defined.

When you create a scope it copies the current global configuration in a new scope and it modifies that copy:

```csharp
using (JsConfig.With(emitLowercaseUnderscoreNames: true, 
                     emitCamelCaseNames: false, 
                     propertyConvention: PropertyConvention.Lenient))
{
    return text.FromJson<T>();
}
```

Scopes weren't mutating global configuration before either, the difference now is that global configuration is never read in a config scope, just its own scoped configuration snapshot and if the global configuration stays immutable there shouldn't be any invalid shadowed copies of static global configuration in local Thread storage (if that was the cause).

The preferred way to set global configuration is to now use `JsConfig.Init()`, e.g:

```csharp
JsConfig.Init(new Config {
    DateHandler = DateHandler.ISO8601,
    AlwaysUseUtc = true,
    EmitLowercaseUnderscoreNames = false,
    EmitCamelCaseNames = false,
    ExcludeDefaultValues = true,                
});
```

You can still `JsConfig.EmitLowercaseUnderscoreNames = true` but using `Init()` will assert that the Global Configuration is not mutated after `Init()` is called.

Also you can assert multiple Init()'s are not called by enabling [StrictMode](http://docs.servicestack.net/debugging#strictmode):

```csharp
SetConfig(new HostConfig { StrictMode = true });
```

Which is on by default during Debug and off by default in Release builds.

This is the most I can do without a repro. An undisposed scope can still cause this issue in the Thread where the scope was created. Otherwise the only other cause I can think of is if the config was cached in a delegate cache but [WriteType](https://github.com/ServiceStack/ServiceStack.Text/blob/master/src/ServiceStack.Text/Common/WriteType.cs) contains the only 2 places where it's used which never caches the value and calls the `GetPropertyName()` method each time.

This change is available from v5.2.1 that's now [available on MyGet](http://docs.servicestack.net/myget).
